packages/core/helper-plugin/src/features/CustomFields.tsx

```diff
@@ -0,0 +1,129 @@
+import * as React from 'react';
+
+import { MessageDescriptor } from 'react-intl';
+
+/**
+ * @TODO: Custom field types should be defined and exported from the registry:
+ * packages/core/admin/admin/src/core/apis/CustomFields.js
+ */
+interface CustomFieldComponents {
+  Input: () => Promise<React.ComponentType>;
+}
+
+type CustomFieldTypes =
+  | 'biginteger'
+  | 'boolean'
+  | 'date'
+  | 'datetime'
+  | 'decimal'
+  | 'email'
+  | 'enumeration'
+  | 'float'
+  | 'integer'
+  | 'json'
+  | 'password'
+  | 'richtext'
+  | 'string'
+  | 'text'
+  | 'time'
+  | 'uid';
+
+type CustomFieldOptionInputs =
+  | 'text'
+  | 'checkbox'
+  | 'checkbox-with-number-field'
+  | 'select-default-boolean'
+  | 'date'
+  | 'select'
+  | 'number'
+  | 'boolean-radio-group'
+  | 'select-date'
+  | 'text-area-enum'
+  | 'select-number'
+  | 'radio-group';
+
+type CustomFieldOptionNames =
+  | 'min'
+  | 'minLength'
+  | 'max'
+  | 'maxLength'
+  | 'required'
+  | 'regex'
+  | 'enum'
+  | 'unique'
+  | 'private'
+  | 'default';
+
+interface CustomFieldOption {
+  intlLabel: MessageDescriptor;
+  description: MessageDescriptor;
+  name: CustomFieldOptionNames;
+  type: CustomFieldOptionInputs;
+  defaultValue?: string | number | boolean | Date;
+}
+
+interface CustomFieldOptionSection {
+  sectionTitle: MessageDescriptor | null;
+  items: CustomFieldOption[];
+}
+
+interface CustomFieldOptions {
+  base?: (CustomFieldOptionSection | CustomFieldOption)[];
+  advanced?: (CustomFieldOptionSection | CustomFieldOption)[];
+  validator?: () => object;
+}
+
+interface CustomField {
+  name: string;
+  pluginId?: string;
+  type: CustomFieldTypes;
+  intlLabel: MessageDescriptor;
+  intlDescription: MessageDescriptor;
+  icon?: React.ComponentType;
+  components: CustomFieldComponents;
+  options?: CustomFieldOptions;
+}
+
+/* -------------------------------------------------------------------------------------------------
+ * Context
+ * -----------------------------------------------------------------------------------------------*/
+
+interface CustomFieldsContextValue {
+  get: (uid: string) => CustomField | undefined;
+  getAll: () => Record<string, CustomField>;
+}
```

Comment:
it would be amazing if one day we can actually create a union of types based on the plugins installed in your strapi app 🤩